### PR TITLE
feat: check for possible duplicated contacts of a person

### PR DESCRIPTION
**Pull request by MaiLinhGroup**
Add new functionality to check for possible duplicated contacts of a person using a strategy like comparing phonenumbers to SDK for clients to use.

Co-authored-by: Mai Linh <mailinh@hello-charles.com>
Co-authored-by: Duc Anh Trinh <duc@hello-charles.com>

**TODO**:
- [x] squash commit messages

### CODE REVIEW

File: src/entities/person/person.ts [hunk 1]
```diff
@@ -333,6 +332,7 @@ export class Person extends UniverseEntity<PersonPayload, PersonRawPayload> {
   public analytics?: PersonPayload['analytics']
   public defaultAddress?: PersonPayload['defaultAddress']
   public languagePreference?: PersonPayload['languagePreference']
+  // public possibleDuplicates?: PersonPayload['possibleDuplicates']
```
Comment (datrinh) [inline src/entities/person/person.ts:335]:
```suggestion
```
Sorry, my dead code :)

File: src/entities/person/possible-duplicates.ts [hunk 2]
```diff
@@ -0,0 +1,82 @@
+import { Universe } from 'src/universe'
+import { EntityRawPayload, UniverseEntityOptions } from '../_base'
+import { PersonRawPayload, PersonPhonenumberRawPayload } from './person'
+import { ChannelUserRawPayload } from './channel-user'
+import { BaseError } from '../../errors'
+
+export interface PossibleDuplicatesRawPayload extends EntityRawPayload{
```
Comment (datrinh) [inline src/entities/person/possible-duplicates.ts:7]:
```suggestion
export interface PossibleDuplicatesRawPayload extends EntityRawPayload {
```

File: src/entities/person/possible-duplicates.ts [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+export class PossibleDuplication {
+  protected universe: Universe
+  protected apiCarrier: Universe
+  protected http: Universe['http']
+  protected options: PossibleDuplicationOptions
+  public initialized: boolean
+  public endpoint: string
+
+  public id?: string
```
Comment (datrinh) [inline src/entities/person/possible-duplicates.ts:40]: Is this actually really optional?

File: src/entities/person/possible-duplicates.ts [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+  constructor (options: PossibleDuplicationOptions) {
+    this.universe = options.universe
+    this.apiCarrier = options.universe
+    this.http = options.http
+    this.options = options
+    this.initialized = options.initialized ?? false
+    this.endpoint = ''
+    this.personId = options.rawPayload?.person_id
+    this.strategies = options.rawPayload?.strategies
```
Comment (datrinh) [inline src/entities/person/possible-duplicates.ts:52]:
```suggestion
    this.personId = options.rawPayload.person_id
    this.strategies = options.rawPayload.strategies
```
I think this can be omitted now after our changes :)

File: src/entities/person/possible-duplicates.ts [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+  public personId: PossibleDuplicatesRawPayload['person_id']
+  public strategies: PossibleDuplicatesRawPayload['strategies']
+
+  constructor (options: PossibleDuplicationOptions) {
+    this.universe = options.universe
+    this.apiCarrier = options.universe
+    this.http = options.http
+    this.options = options
+    this.initialized = options.initialized ?? false
+    this.endpoint = ''
```
Comment (datrinh) [inline src/entities/person/possible-duplicates.ts:50]: We seem to never set the endpoint anywhere anymore. How does it work then? 😅

Comment (MaiLinhGroup) [reply src/entities/person/possible-duplicates.ts:50]: we do actually, but in `people.checkDuplicates()` instead and here is my thinking why and please comment and correct me if I'm wrong:
- we cannot set the endpoint in the constructor of `PossibleDuplication` because it will be instantiate after the call when the endpoint returns it, so the moment we call `{{HOST}}/api/v0/people/:id/check/duplicates` it won't be there yet
- but `people` will so `this.endpoint` will be what is set in its constructor (`this.endpoint = 'api/v0/people'`)
- we complete the endpoint then in the aforementioned `people.checkDuplicates()`
- I thought this could be why after Nick showed me how to test our endpoint in the `agent-ui/src/pages/dev/Sandbox.vue/mounted()::his.$universe.person({ id: '2c26b6f1-3a45-4b94-ac03-553d8e7ed832' }).checkDuplicates().then(res => { console.log(res) })` <- you can see how we pass the person id to instantiate the person we call the `checkDuplicates()` method for
WDYT?

File: src/entities/person/possible-duplicates.ts (hunk 3 above)
Comment (MaiLinhGroup) [reply src/entities/person/possible-duplicates.ts:40]: Right now it is not part of the mock data we're returning on the endpoint indeed :), this `id` will be one of the entries of `people.possible_duplicated_contacts_lists_items`. For now I don't return the raw entry/item just as-is from the table but the return will be a composite/results of some queries over this and other tables as well so I don't see a need right now to send the id's in the return but this might change in the future when we may have some use case where you would need to load a specific item from `people.possible_duplicated_contacts_lists_items` for a person without rechecking and therefore re-insert the entries for that person :)

TLDR; this is for now indeed optional and might change in the future then we can be more strict and remove the optional

**Review by eljefedelrodeodeljefe — CHANGES_REQUESTED** [review]
Remove everything from /dist pls

File: src/entities/person/person.ts [hunk 6]
```diff
@@ -961,6 +961,20 @@ export class Person extends UniverseEntity<PersonPayload, PersonRawPayload> {
...
     }
   }
+
+  public async checkDuplicates (): Promise<PossibleDuplicatesRawPayload[]> {
+    if (this.id === null || this.id === undefined) throw new TypeError('people checkDuplicates requires id to be set.')
+    try {
+      const res = await this.http?.getClient().put<PossibleDuplicatesRawPayload[]>(
+        `${this.universe?.universeBase}/${this.endpoint}/${this.id}/check/duplicates`,
+        null,
+        { params: { strategies: ['global_phonenumber'] } })
```
Comment (eljefedelrodeodeljefe) [inline src/entities/person/person.ts:971]: i'd prefer the UI to hardcode this

File: src/entities/person/person.ts [hunk 7]
```diff
@@ -961,6 +961,20 @@ export class Person extends UniverseEntity<PersonPayload, PersonRawPayload> {
       throw new PersonPreviewNotificationError(undefined, { error: err })
     }
   }
+
+  public async checkDuplicates (): Promise<PossibleDuplicatesRawPayload[]> {
```
Comment (eljefedelrodeodeljefe) [inline src/entities/person/person.ts:965]: for that case we'd need need to define an interface for the query that allows for
```
{
  strategies: ['global_phonenumber']
}
```

which then should be passed to...

File: src/entities/person/person.ts [hunk 8]
```diff
@@ -961,6 +961,20 @@ export class Person extends UniverseEntity<PersonPayload, PersonRawPayload> {
...
+    if (this.id === null || this.id === undefined) throw new TypeError('people checkDuplicates requires id to be set.')
+    try {
+      const res = await this.http?.getClient().put<PossibleDuplicatesRawPayload[]>(
+        `${this.universe?.universeBase}/${this.endpoint}/${this.id}/check/duplicates`,
```
Comment (eljefedelrodeodeljefe) [inline src/entities/person/person.ts:969]:
```
`${this.universe?.universeBase}/${this.endpoint}/${this.id}/check/duplicates{qs.stringify(nameOfParamDefinedAbove ?? {}, {addQueryPrefix: true })}`,
```

Comment (eljefedelrodeodeljefe) [reply src/entities/person/person.ts:969]: ...see other querystring parsing locations

File: src/entities/person/person.ts (hunk 6 above)
Comment (eljefedelrodeodeljefe) [reply src/entities/person/person.ts:971]: ...also this would be removed, see comments above

File: src/entities/person/possible-duplicates.ts [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+  person: Pick<PersonRawPayload, 'id'| 'kind' | 'name'>
+  channel_user?: Pick<ChannelUserRawPayload, 'id' | 'name' | 'phone'>
+  phone_number?: Pick<PersonPhonenumberRawPayload, 'id' | 'value'>
+}
+
+interface PossibleDuplicationOptions extends UniverseEntityOptions {
+  rawPayload: PossibleDuplicatesRawPayload
+}
+
+export class PossibleDuplication {
```
Comment (eljefedelrodeodeljefe) [inline src/entities/person/possible-duplicates.ts:32]: making this a class, instead of interface is kinda optional. Class would only be needed to eventually make subsequent actions on that duplicate. For now that is likely not planned, but a case would be making the merge of the resulting person from this instance. Since the UX flow is rather to open a modal and configure that merge, that case though is not strong

File: src/entities/person/possible-duplicates.ts (hunk 5 above)
Comment (eljefedelrodeodeljefe) [reply src/entities/person/possible-duplicates.ts:50]: This would become unnecessary if my last comment applies.

File: src/entities/person/possible-duplicates.ts [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+  global_phonenumber?: StrategiesResult[]
+  name_exact_matches?: StrategiesResult[]
+}
+
+interface StrategiesResult {
+  document: StrategiesDocument
+  kind: string
+}
+
+interface StrategiesDocument {
```
Comment (eljefedelrodeodeljefe) [inline src/entities/person/possible-duplicates.ts:22]: Could be public, meaning exported, so consumers can import the type

Comment (eljefedelrodeodeljefe) [reply src/entities/person/possible-duplicates.ts:22]: Applies for all the new interfaces

File: src/entities/person/possible-duplicates.ts [hunk 11]
```diff
@@ -0,0 +1,82 @@
+import { Universe } from 'src/universe'
+import { EntityRawPayload, UniverseEntityOptions } from '../_base'
+import { PersonRawPayload, PersonPhonenumberRawPayload } from './person'
+import { ChannelUserRawPayload } from './channel-user'
+import { BaseError } from '../../errors'
+
+export interface PossibleDuplicatesRawPayload extends EntityRawPayload {
+  readonly person_id: string
+  readonly strategies: PossibleDuplicatesStrategies
+}
```
Comment (eljefedelrodeodeljefe) [inline src/entities/person/possible-duplicates.ts:10]: Likely this interface is not aligned with the return data

File: src/entities/person/possible-duplicates.ts [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+      this.deserialize(options.rawPayload)
+    }
+  }
+
+  protected deserialize (rawPayload: PossibleDuplicatesRawPayload): PossibleDuplication {
+    this.id = rawPayload.id
+    this.personId = rawPayload.person_id
+    this.strategies = rawPayload.strategies
+
+    return this
```
Comment (eljefedelrodeodeljefe) [inline src/entities/person/possible-duplicates.ts:64]: Also this does not seem to be aligned with the incoming data, no?

**Review by eljefedelrodeodeljefe — CHANGES_REQUESTED** [review]

**Comment by eljefedelrodeodeljefe** [comment]
For workflow: you can also TDD it against something like https://github.com/c-commerce/charles-browser-sdk/blob/master/test/me/get.test.ts

File: src/entities/person/possible-duplicates.ts (hunk 12 above)
Comment (MaiLinhGroup) [reply src/entities/person/possible-duplicates.ts:64]: see comment above to returned data

File: src/entities/person/possible-duplicates.ts (hunk 11 above)
Comment (MaiLinhGroup) [reply src/entities/person/possible-duplicates.ts:10]: It does actually, the return data in the console in FE (see screenshot) matches with what I'd also get back when using postman to access BE directly, or do I might get you wrong here? then please be so kind to specify :) 

The returned data is currently of form `{person_id: uuid, strategies: { global_phonenumber[], exact_name_matches[] }}` and when we add new strategies like `exact_email_matches` etc. then we can add it as a new property of type `StrategiesResult` to `PossibleDuplicatesStrategies` and it will have `document` and `kind` as properties like the others. Maybe `StrategiesDocument` needs to be extended then.

Screenshot 2021-09-13 at 20 06 48

File: src/entities/person/possible-duplicates.ts (hunk 9 above)
Comment (MaiLinhGroup) [reply src/entities/person/possible-duplicates.ts:32]: 👍 ok let's make it an interface then as I don't have strong argument either for it to be a class

File: src/entities/person/possible-duplicates.ts (hunk 11 above)
Comment (MaiLinhGroup) [reply src/entities/person/possible-duplicates.ts:10]: I think I might get now what you were hinting or questioning? in the previous version the checkDuplicates was wrongly assuming it would return `PossibleDuplicatesRawPayload[]` which the API isn't! in the return data (`data[0]`) is instead only the object `PossibleDuplicatesRawPayload` that has a property `strategies` which again has the list with the duplicated contacts!

later when we want to return for all people each their duplicates then we actually would have `PossibleDuplicatesRawPayload[]` but for now this was about checking duplicates for one person. WDYT?

**Comment by MaiLinhGroup** [comment]
I give it a try this weekend to see if we can get rid of the dist folder somehow or not 🤔

**Review by datrinh — APPROVED** [review]
LGTM
